File: app/src/main/java/com/example/exitpro/Activity/LoginActivity.java
package com.example.exitpro.Activity;

import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.FragmentTransaction;

import android.app.ProgressDialog;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Toast;

import com.android.volley.Request;
import com.android.volley.RequestQueue;
import com.android.volley.Response;
import com.android.volley.VolleyError;
import com.android.volley.toolbox.JsonObjectRequest;
import com.android.volley.toolbox.Volley;
import com.example.exitpro.Config.Config;
import com.example.exitpro.Fragment.OTPVerification;
import com.example.exitpro.R;

import org.json.JSONException;
import org.json.JSONObject;

public class LoginActivity extends AppCompatActivity {

    // UI elements
    private Button loginBtn;
    private EditText guardID;
    private ProgressDialog progressDialog;
    private RequestQueue requestQueue;

    // URL for login request
    private static final String LOGIN_URL = Config.BASE_URL + "/security/login";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);

        // Initialize UI elements
        loginBtn = findViewById(R.id.login_button);
        guardID = findViewById(R.id.guard_id_input);

        // Initialize Volley request queue
        requestQueue = Volley.newRequestQueue(this);

        // Set click listener for the login button
        loginBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                // Get guard ID from EditText
                String guardId = guardID.getText().toString();

                // Validate guard ID (if necessary, add your own validation logic)

                // Send guard ID to the backend
                sendGuardID(guardId);
            }
        });
    }

    /**
     * Sends the guard ID to the backend server.
     *
     * @param guardId The guard ID to send.
     */
    private void sendGuardID(String guardId) {
        // Show loading dialog
        showLoadingDialog();

        // Create JSON object with guard ID
        JSONObject jsonBody = new JSONObject();
        try {
            jsonBody.put("guardId", guardId);
        } catch (JSONException e) {
            e.printStackTrace();
        }

        // Create a request to send the guard ID
        JsonObjectRequest request = new JsonObjectRequest(
                Request.Method.PUT,
                LOGIN_URL,
                jsonBody,
                new Response.Listener<JSONObject>() {
                    @Override
                    public void onResponse(JSONObject response) {
                        // Dismiss loading dialog
                        dismissLoadingDialog();
                        try {
                            boolean isSuccess = response.getBoolean("isSuccess");
                            if (isSuccess) {
                                // If login is successful, start OTP verification fragment
                                otpVerificationFragment();
                            } else {
                                // If login fails, show error
                                dismissLoadingDialog();
                                guardID.setError("Wrong credentials");
                            }
                        } catch (JSONException e) {
                            e.printStackTrace();
                        }
                    }
                },
                new Response.ErrorListener() {
                    @Override
                    public void onErrorResponse(VolleyError error) {
                        // Dismiss loading dialog and show error message
                        dismissLoadingDialog();
                        Toast.makeText(LoginActivity.this, "Error: " + error.getMessage(), Toast.LENGTH_SHORT).show();
                    }
                }
        );

        // Add the request to the request queue
        requestQueue.add(request);
    }

    /**
     * Starts the OTP verification fragment.
     */
    private void otpVerificationFragment() {
        // Create a bundle to pass guard ID to the fragment
        Bundle bundle = new Bundle();
        bundle.putString("Guard ID", guardID.getText().toString());

        // Create and set up the OTP verification fragment
        OTPVerification otpVerificationFragment = new OTPVerification();
        otpVerificationFragment.setArguments(bundle);

        // Begin a fragment transaction to replace the current fragment with the OTP verification fragment
        FragmentTransaction transaction = getSupportFragmentManager().beginTransaction();
        transaction.replace(R.id.otpVerificationFragment, otpVerificationFragment);
        transaction.addToBackStack(null);
        transaction.commit();
    }

    /**
     * Shows the loading dialog.
     */
    private void showLoadingDialog() {
        progressDialog = new ProgressDialog(LoginActivity.this);
        progressDialog.setCancelable(false);
        progressDialog.setMessage("Sending OTP...");
        progressDialog.show();
    }

    /**
     * Dismisses the loading dialog if it is showing.
     */
    private void dismissLoadingDialog() {
        if (progressDialog != null && progressDialog.isShowing()) {
            progressDialog.dismiss();
        }
    }
}

File: app/src/main/java/com/example/exitpro/activity/LoginActivity.kt
package com.example.exitpro.activity

import android.app.Dialog
import android.os.Bundle
import android.view.Window
import android.widget.Button
import android.widget.EditText
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.FragmentTransaction
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.lifecycle.Lifecycle
import com.example.exitpro.R
import com.example.exitpro.data.api.RetrofitClient
import com.example.exitpro.data.repository.ExitProRepository
import com.example.exitpro.fragment.OTPVerification
import com.example.exitpro.viewmodel.LoginViewModel
import com.example.exitpro.viewmodel.ViewModelFactory
import com.example.exitpro.viewmodel.state.LoginUiState
import kotlinx.coroutines.launch

class LoginActivity : AppCompatActivity() {
    private lateinit var loginBtn: Button
    private lateinit var guardID: EditText
    private var loadingDialog: Dialog? = null
    
    // MVVM - ViewModel
    private val viewModel: LoginViewModel by lazy {
        val repository = ExitProRepository(RetrofitClient.apiService)
        val factory = ViewModelFactory(repository)
        factory.create(LoginViewModel::class.java)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_login)

        // Initialize UI elements
        loginBtn = findViewById(R.id.login_button)
        guardID = findViewById(R.id.guard_id_input)

        // Set click listener for the login button
        loginBtn.setOnClickListener {
            val guardId = guardID.text.toString().trim()

            if (guardId.isNotEmpty()) {
                sendGuardID(guardId)
            } else {
                guardID.error = "Guard ID cannot be empty"
            }
        }
        
        // Observe ViewModel state
        observeLoginState()
    }

    /**
     * Send guard ID to server for login using ViewModel (MVVM pattern).
     */
    private fun sendGuardID(guardId: String) {
        viewModel.login(guardId)
    }
    
    /**
     * Observe login state from ViewModel and update UI accordingly.
     */
    private fun observeLoginState() {
        lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                viewModel.loginState.collect { state ->
                    when (state) {
                        is LoginUiState.Idle -> {
                            // Do nothing
                        }
                        is LoginUiState.Loading -> {
                            showLoadingDialog()
                        }
                        is LoginUiState.Success -> {
                            dismissLoadingDialog()
                            // Get the guard ID from the EditText to pass to OTP fragment
                            val guardId = guardID.text.toString().trim()
                            otpVerificationFragment(guardId)
                            viewModel.resetLoginState()
                        }
                        is LoginUiState.Error -> {
                            dismissLoadingDialog()
                            guardID.error = state.message
                            Toast.makeText(this@LoginActivity, state.message, Toast.LENGTH_SHORT).show()
                            viewModel.resetLoginState()
                        }
                    }
                }
            }
        }
    }

    private fun otpVerificationFragment(guardId: String) {
        val bundle = Bundle().apply {
            putString("Guard ID", guardId)
        }

        val otpVerificationFragment = OTPVerification().apply {
            arguments = bundle
        }

        supportFragmentManager.beginTransaction().apply {
            replace(R.id.otpVerificationFragment, otpVerificationFragment)
            addToBackStack(null)
            setTransition(FragmentTransaction.TRANSIT_FRAGMENT_FADE)
            commit()
        }
    }

    /**
     * Display a loading dialog with sending OTP message.
     */
    private fun showLoadingDialog() {
        loadingDialog = Dialog(this).apply {
            requestWindowFeature(Window.FEATURE_NO_TITLE)
            setContentView(R.layout.loading_dialog)
            setCancelable(false)
            window?.setBackgroundDrawableResource(android.R.color.transparent)
            show()
        }
    }

    /**
     * Dismiss the loading dialog if currently showing.
     */
    private fun dismissLoadingDialog() {
        loadingDialog?.takeIf { it.isShowing }?.dismiss()
    }

}
